### ISSUE TITLE
Add data as demonstrator/scheduled sampling. 

### ISSUE THREAD

**Issue by PetrochukM** [OpenNMT/OpenNMT-py#113]
"Teacher forcing" is the concept of using the real target outputs as each next input, instead of using the decoder's guess as the next input. Using teacher forcing causes it to converge faster.

**Comment by jekbradbury** [comment]
Everything in OpenNMT currently uses teacher forcing for training.

**Comment by Deepblue129** [comment]
@jekbradbury Having a hard time finding the teacher forcing inside the models. 

Found a parameter called `input_feed` I also see that the Decoder module model accepts `input`. For Transformer model, I see that the input is added in https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/Models.py#L251. Looks like `state.previous_input` is used to help that.  Is `input_feed` related to teacher forcing?

For other NMT models that are not transformer, I do not see where teacher forcing is used.

**Comment by jekbradbury** [comment]
Here https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/Models.py#L227 `input` is the real (ground-truth) target outputs, shifted one timestep over. That's called from https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/Models.py#L379, which does the shifting.

**Comment by jekbradbury** [comment]
I would recommend reading the key papers that this codebase implements, in particular https://arxiv.org/abs/1508.04025, as well as the OpenNMT paper (https://arxiv.org/abs/1701.02810, although this describes the Lua version).

**Comment by falcondai** [comment]
@Deepblue129 are you asking for support for [scheduled sampling](https://arxiv.org/abs/1506.03099)?

**Comment by Deepblue129** [comment]
@jekbradbury Thanks for the code lines. I found that the input is a parameter to the decoder. Though I did not see it used for teacher forcing. 

@falcondai Yup! 

Found a summary of the paper that nicely summarizes the feature:
'So the authors propose a different training procedure, where at training time each *conditioning* ground truth token is sometimes replaced by the model's previous prediction. The choice of replacing the ground truth by the model's prediction is made by "flipping a coin" with some probability, independently for each token. Importantly, the authors propose to start with a high probability of using the ground truth (i.e. start close to ML) and anneal that probability closer to 0, according to some schedule (thus the name Schedule Sampling). '
https://www.evernote.com/shard/s189/sh/c9ac2e3f-a150-4d0c-9a44-16657e5d42cd/5eb49d50695c903ca1b4a04934e63363

**Comment by quanpn90** [comment]
Maybe you got some confusion between Teacher Forcing (using reference labels as inputs) and Free Running (using sampled labels as inputs). Basically you'd want to compute the probability distribution, then sample a label from it (or take the argmax) and use it as inputs for the next step. 

Input-Feeding (Adding the previous hidden state to current input) is irrelevant to Teacher-Forcing or Free-Running. Current project is using Teacher Forcing + Input Feeding.

**Comment by Deepblue129** [comment]
@quanpn90 Teacher forcing is when the target sequence is fed back into the model.

Similar to https://github.com/IBM/pytorch-seq2seq/blob/master/seq2seq/models/DecoderRNN.py#L154 

Are we talking about the same thing?

**Comment by quanpn90** [comment]
Yes, I think we are. Your 'target sequence' is what I meant by 'reference labels'. 

Correct me if i'm wrong, but I learned that from here: 

https://papers.nips.cc/paper/6099-professor-forcing-a-new-algorithm-for-training-recurrent-networks.pdf

"The Teacher Forcing algorithm trains recurrent networks by supplying observed sequence values as inputs during training" which is what OpenNMT-py is doing here.

**Comment by Deepblue129** [comment]
@quanpn90 Okay! Happy to hear OpenNMT-py does this. Can you point me to where it does? I could not find the line of code where the observed sequence value is fed as an input during training.

**Comment by quanpn90** [comment]
:D The "observed sequence values" are the target sentences themselves, which are the 'tgt' in NMTModel.forward() function.

**Comment by srush** [comment]
Short answer, this is a bit hard to implement in the current version since it breaks parallelism. But I think it is important. We will consider ways to make this happen.

**Comment by ibulu** [comment]
This would be a valuable feature. I have implemented poor man's version with the inputfeeddecoder. There are some interesting choices to be made once you disable teacher forcing. For instance, do you exclude padding idx from loss calculation? If one excludes padding idx from loss calculation, than the resulting value underestimates the loss.

**Comment by vince62s** [comment]
closing this, adding in the projects.
